torch/testing/_comparison.py

```diff
@@ -949,13 +955,13 @@ def _compare_sparse_compressed_values(
             )
 
         self._compare_regular_values_equal(
-            compressed_indices_method(actual),
-            compressed_indices_method(expected),
+            compressed_indices_method(actual).to(torch.int64),
+            compressed_indices_method(expected).to(torch.int64),
```

Comment:
Compressed and plain indices in the CSR / CSC / BSR / BSC can be `torch.int32` _or_ `torch.int64`. While the same dtype is enforced within a single tensor, it can be different for two tensors. Thus, we need to convert them to the same dtype here, or `isclose` bail out.

This behavior is the same as before where all integer comparisons where upcasted to int64, but now we need to do it explicitly. Ideally, we wouldn't do that if both inputs are int32, but if it was not an issue before, I prefer to keep the implementation simple here.

Reply:
Considering the aim of this PR, that is, to reduce the memory footprint, it might be worth finding the maximal dtype of indices, say:
```python
        indices_dtype = compressed_indices_method(actual).dtype
        expected_indices_dtype = compressed_indices_method(expected).dtype
        indices_dtype = indices_dtype if indices_dtype == expected_indices_dtype else torch.int64
        self._compare_regular_values_equal(
            compressed_indices_method(actual).to(indices_dtype),
            compressed_indices_method(expected).to(indices_dtype),
            identifier=f"Sparse {format_name} {compressed_indices_method.__name__}",
        )
```

Reply:
The found `indices_dtype` can be used for `plain_indices` as well.